src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -161,45 +227,122 @@ double LMNNFunction<MetricType>::Evaluate(const arma::mat& transformation,
     {
       // Calculate cost due to distance between target neighbors & data point.
       double eval = metric.Evaluate(transformedDataset.col(i),
-                        transformedDataset.col(targetNeighbors(j, i)));
+                          transformedDataset.col(targetNeighbors(j, i)));
       cost += (1 - regularization) * eval;
     }
 
+    // Calculate norm of change in transformation.
+    double transformationDiff = 0;
+    if (arma::is_finite(lastTransformationIndices(i)))
+    {
+      transformationDiff = arma::norm(transformation -
+          oldTransformationMatrices[lastTransformationIndices(i)]);
+    }
```

Comment:
I think it is possible to do this for amsgrad and sgd (change can be made in EvaluatewithGradient()) but will have to keep it like this for bbsgd.

Reply:
Right, you can leave it like this and I think it's fine.  What I was thinking was more along the lines of something like this:

```
std::set<size_t> uniqueIndices;
std::map<size_t, double> transformationDiffs;
for (size_t i = begin; i < begin + batchSize; ++i)
{
  if (transformationDiffs.count(lastTransformationIndices[i]) == 0)
  {
    // If you make the change so that oldTransformationMatrices[0] is empty, this will work.
    if (lastTransformationIndices[i] == 0)
    {
      transformationDiffs[0] = 0.0; // This won't be used anyway...
    }
    else
    {
      transformationDiffs[lastTrasformationIndices[i]] = arma::norm(transformation -
          oldTransformationMatrices[lastTransformationIndices(i)]);
    }
  }
}
```

I think that will work, but I haven't tried or debugged it.  Then, later, instead of using `transformationDiff` for the bound calculation, you'd use `transformationDiff[oldTransformationIndices[i]]`.